aten/src/ATen/native/Sorting.cpp

```diff
@@ -201,6 +202,90 @@ std::tuple<Tensor, Tensor> kthvalue(
   return std::make_tuple(values, indices);
 }
 
+std::tuple<Tensor&, Tensor&> topk_out_cpu(
+    Tensor& values,
+    Tensor& indices,
+    const Tensor& self,
+    int64_t k,
+    int64_t dim_,
+    bool largest,
+    bool sorted) {
+  int64_t dim = maybe_wrap_dim(dim_, self.dim(), /*wrap_scalar=*/true);
+  AT_CHECK(
+      k >= 0 && k <= (self.dim() > 0 ? self.size(dim) : 1),
+      "selected index k out of range");
+
+  _allocate_or_resize_output_with_indices(values, indices, self, dim_, k);
+  if (self.dim() == 0 && self.numel() == 1) {
+    values.copy_(self);
+    indices.zero_();
+    return std::forward_as_tuple(values, indices);
+  }
+  auto tmp_values = self.clone();
+  AT_DISPATCH_ALL_TYPES(self.scalar_type(), "topk_cpu", [&] {
+    dim_apply(
```

Comment:
This looks very close to TensorIterators. In fact, dim_apply overall looks like it could be replaced by that. This would yield much cleaner and usually faster code. @VitalyFedyunin, please provide guidance.

Reply:
I believe the reason why TensorIterator cant be used here is that they cant take a set of input values and return a single output value (which is the case in sort / topk). But it would be great to extend TensorIterator to support this use-case